nix/haskell-project.nix

```diff
@@ -55,14 +58,31 @@ in
                   hpack
                 ''
             else root;
+          setSrc = name: root: pkg: hlib.overrideSrc {
+                src =
+                  lib.cleanSourceWith {
+                    name = "source-${name}-${pkg.version}";
+                    src =
+                      # NOTE: Even though cabal2nix does run hpack automatically,
+                      # buildFromCabalSdist does not. So we must run hpack ourselves at
+                      # the original source level.
+                      realiseHpack name root;
+                  };
+              } pkg;
         in
         lib.mapAttrs
-          (name: value:
-            # NOTE: Even though cabal2nix does run hpack automatically,
-            # buildFromCabalSdist does not. So we must run hpack ourselves at
-            # the original source level.
-            let root = realiseHpack name value.root;
-            in fromSdist (self.callCabal2nix name root { })
+          (name: pkgCfg:
+            lib.pipe pkgCfg
+            [
+              (pkgCfg: self.callCabal2nix name pkgCfg.root { })
+
+              # Avoid rebuilding because of changes in parent directories
+              (setSrc name pkgCfg.root)
```

Comment:
What's the rationale behind the use of `lib.cleanSourceWith` somehow not triggering rebuilds that are otherwise triggered?

Reply:
You may want to add a test to reproduce the rebuild such that only the use of `setSrc` can fix it.

Reply:
> I removed the `srcSrc` call, but the tests still pass (on macOS at least)

Bizarrely, this is not the case with haskell-multi-nix. Your PR *does* prevent rebuild. Using `master` branch of haskell-flake causes rebuilds when README.md changes. What's different about `test/with-subdir` though?

Reply:
Looking inside the two `.drv`, I see that the source hash changes, yet the tarball contents themselves are identical:

```
❯ nix run nixpkgs#pkgdiff /nix/store/9qy8hbsbawcw1crmr9nbljr3yyn5skzw-source.tar.gz /nix/store/f2daz0iv6dn8j6iw95s16rw5zn27ca8w-source.tar.gz
reading packages ...
WARNING: different package names ("9qy8hbsbawcw1crmr9nbljr3yyn5skzw" and "f2daz0iv6dn8j6iw95s16rw5zn27ca8w")
comparing packages ...
creating report ...
result: UNCHANGED
report: pkgdiff_reports/9qy8hbsbawcw1crmr9nbljr3yyn5skzw/source_to_source/changes_report.html
```

Reply:
> What's the rationale behind the use of `lib.cleanSourceWith` somehow not triggering rebuilds that are otherwise triggered?

`pkgCfg.root` is a value like `"${root}/mypackage"`, so something like `"/nix/store/...-source/mypackage"`, where `/nix/store/...-source` contains everything at the root of the flake.

`cleanSourceWith` does not copy things from the parent directory, so it returns a store path that only contains the relevant directory.

> Looking inside the two `.drv`, I see that the source hash changes, yet the tarball contents themselves are identical:

That means that `cabal sdist` is doing its job. You'll need the `src.src` - the source of the sdist build - to be identical too. Otherwise, the sdist builds are going to be different.

Reply:
Thanks; added this in 600a179